File: exercism/haskell/kindergarten-garden/src/Garden.hs
module Garden (garden, defaultGarden, lookupPlants, Plant(..)) where

import Data.List (sort, transpose)
import Data.List.Split (chunksOf)

data Plant = Radishes
           | Violets
           | Clover
           | Grass
  deriving (Eq, Show)

type Child = String

defaultChildren :: [Child]
defaultChildren = ["Alice", "Bob", "Charlie", "David",
                   "Eve", "Fred", "Ginny", "Harriet",
                   "Ileana", "Joseph", "Kincaid", "Larry"]

plantOfChar :: Char -> Plant
plantOfChar 'R' = Radishes
plantOfChar 'V' = Violets
plantOfChar 'C' = Clover
plantOfChar 'G' = Grass
plantOfChar _ = error "Unknown plant in kindergarten garden!"

defaultGarden :: String -> [(Child, [Plant])]
defaultGarden = garden defaultChildren

garden :: [Child] -> String -> [(Child, [Plant])]
garden children cells = zip (sort children) (plants cells)
  where
    plants = map (map plantOfChar) . parcells
    parcells = map concat . transpose . map (chunksOf 2) . lines

lookupPlants :: Child -> [(Child, [Plant])] -> [Plant]
lookupPlants child = snd . head . filter (\(c, _) -> c == child)

File: exercism/elixir/kindergarten-garden/garden.exs
defmodule Garden do
  @doc """
    Accepts a string representing the arrangement of cups on a windowsill and a
    list with names of students in the class. The student names list does not
    have to be in alphabetical order.

    It decodes that string into the various gardens for each student and returns
    that information in a map.
  """

    @students [ 
      :alice, :bob, :charlie, :david,
      :eve, :fred, :ginny, :harriet,
      :ileana, :joseph, :kincaid, :larry,
    ]

    @plants %{
      "C" => :clover,
      "G" => :grass,
      "R" => :radishes,
      "V" => :violets,
    }
  
  @spec info(String.t(), list) :: map
  def info(info_string, student_names \\ @students) do
    student_names = Enum.sort(student_names)

    map = Enum.reduce(student_names, %{}, &(Map.put(&2, &1, {})))

    rows = String.split(info_string, "\n")
    rows = Enum.map(rows, &(
      &1 |> String.graphemes |> Enum.chunk(2)
    ))

    # My elixir version doesn't have Enum.zip/1 yet :-(
    Enum.zip(Enum.at(rows, 0), Enum.at(rows, 1))
    |> Enum.zip(student_names)
    |> Enum.reduce(map, &insert/2)
  end

  defp insert({{row1, row2}, student}, map) do
    Enum.reduce(row1 ++ row2, map, fn(field, map) ->
      Map.update!(map, student, &(Tuple.append(&1, @plants[field])))
    end)
  end
end
